### PR TITLE
Prioritized Task Scheduling API

### PR DESCRIPTION
**Pull request by hamishwillee**
FF101 delivers the [Prioritized Task Scheduling API](https://wicg.github.io/scheduling-apis/) (a prototype) in https://bugzilla.mozilla.org/show_bug.cgi?id=1734997

This documents the API. Guide is in the API Overview doc.
It also includes update to experimental features.

Other docs work can be tracked in #15468

### CODE REVIEW

File: files/en-us/web/api/taskprioritychangeevent/taskprioritychangeevent/index.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+new TaskPriorityChangeEvent(type, options);
+```
+
+### Parameters
+
+- `type`
+  - : A string with the case-sensitive name of the associated event.
+    User agents set it to `prioritychange`.
+
+- `options` {{optional_inline}} <!-- required? looks like it from IDL but isn't there also a default? -->
```
Comment (hamishwillee) [inline files/en-us/web/api/taskprioritychangeevent/taskprioritychangeevent/index.md:32]: The IDL for this [in spec](https://wicg.github.io/scheduling-apis/#sec-task-priority-change-event) says:
```
[Exposed=(Window, Worker)]
interface TaskPriorityChangeEvent : Event {
  constructor(DOMString type, TaskPriorityChangeEventInit priorityChangeEventInitDict);

  readonly attribute TaskPriority previousPriority;
};

dictionary TaskPriorityChangeEventInit : EventInit {
  required TaskPriority previousPriority;
};
```
Since the `previousPriority` is marked as **required** in the dictionary, does that mean the priorityChangeEventInitDict is also required?

File: files/en-us/web/api/taskprioritychangeevent/previousPriority/index.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+## Value
+
+A string, indicating the associated task's priority before it was changed.
+This will be one of: [`"user-blocking"`](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#user-blocking), [`"user-visible"`](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#user-visible), [`"background"`](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#background).
+
+<!-- 
+## Examples
+
+Maybe show triggering and reading the previous an new event
```
Comment (hamishwillee) [inline files/en-us/web/api/taskprioritychangeevent/previousPriority/index.md:27]: TODO: Example

File: files/en-us/web/api/taskprioritychangeevent/index.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+- {{domxref("TaskPriorityChangeEvent.previousPriority")}} {{readonlyInline}}
+  - : Returns the [priority](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#task_priorities) of the corresponding {{domxref("TaskSignal")}} _before_ this [`prioritychange`](/en-US/docs/Web/API/prioritychange_event) event.
+
+## Methods
+
+_This interface has no methods of its own, but inherits the methods of its parent, {{domxref("Event")}}._
+
+
+<!--
```
Comment (hamishwillee) [inline files/en-us/web/api/taskprioritychangeevent/index.md:38]: TODO: Example

File: files/en-us/web/api/prioritychange_event/index.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  - Event
+  - Reference
+  - Prioritized Task Scheduling API
+  - prioritychange
+  - Experimental
+browser-compat: api.TaskSignal.prioritychange_event
+---
+{{APIRef("Prioritized Task Scheduling API")}} {{SeeCompatTable}}
+
+The **`prioritychange`** event is sent to a {{domxref('TaskSignal')}} if its [priority](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#task_priorities) is changed.
```
Comment (hamishwillee) [inline files/en-us/web/api/prioritychange_event/index.md:14]: Is "event is sent to a `TaskSignal`" the correct way of describing this? 

From [the spec](https://wicg.github.io/scheduling-apis/#sec-task-priority-change-event) you have a `TaskController` from which you can get a `TaskSignal` (derived from `AbortController` and `AbortSignal`, prespectively). If you set the priority of the controller, it sets the priority of the signal, which causes `prioritychange` to be dispatched to observers.

The event target is, I believe the `TaskSignal` - ie. that's what observers get.

File: files/en-us/web/api/prioritychange_event/index.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+## Event properties
+
+- {{domxref('TaskPriorityChangeEvent.previousPriority')}}
+  - : Indicates the previous priority of the task (before it was changed).
+    The new/updated priority is read from `event.target.priority` ([`TaskSignal.priority`](/en-US/docs/Web/API/TaskSignal/priority)).
+
+## Usage notes
+
+<!-- TBD -->
```
Comment (hamishwillee) [inline files/en-us/web/api/prioritychange_event/index.md:41]: TBD

File: files/en-us/web/api/prioritychange_event/index.md (hunk 4 above)
Comment (teoli2003) [reply files/en-us/web/api/prioritychange_event/index.md:14]: Looks good to me: sent to or fired at.

File: files/en-us/web/api/prioritychange_event/index.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
+---
+title: prioritychange event
```
Comment (teoli2003) [inline files/en-us/web/api/prioritychange_event/index.md:2]:
```suggestion
title: 'TaskSignal: prioritychange event'
```

File: files/en-us/web/api/prioritychange_event/index.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+slug: Web/API/prioritychange_event
```
Comment (teoli2003) [inline files/en-us/web/api/prioritychange_event/index.md:3]: Should be moved to `Web/API/TaskSignal/prioritychange_event`.

File: files/en-us/web/api/taskprioritychangeevent/taskprioritychangeevent/index.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
+---
+title: TaskPriorityChangeEvent.TaskPriorityChangeEvent()
```
Comment (teoli2003) [inline files/en-us/web/api/taskprioritychangeevent/taskprioritychangeevent/index.md:2]:
```suggestion
title: TaskPriorityChangeEvent()
```

File: files/en-us/web/api/taskprioritychangeevent/taskprioritychangeevent/index.md (hunk 1 above)
Comment (teoli2003) [reply files/en-us/web/api/taskprioritychangeevent/taskprioritychangeevent/index.md:32]: It is mandatory too. If not the webidl would have been:

`constructor(DOMString type, optional TaskPriorityChangeEventInit priorityChangeEventInitDict);`

File: files/jsondata/GroupData.json [hunk 9]
```diff
@@ -1047,6 +1047,14 @@
       ],
       "events": []
     },
+    "Prioritized Task Scheduling API": {
+      "overview": ["Prioritized Task Scheduling API"],
+      "guides": [],
+      "interfaces": ["Scheduler", "TaskController", "TaskSignal", "AbortSignal", "TaskPriorityChangeEvent"],
+      "methods": [],
+      "properties": [],
+      "events": ["WindowOrWorkerGlobalScope: prioritychange_event"]
```
Comment (teoli2003) [inline files/jsondata/GroupData.json:1056]: As it is on TaskSignal (if you follow my comment above), you don't need to mention it there.

**Review by teoli2003 — COMMENTED** [review]
Added a few drive-by comments.

File: files/en-us/web/api/prioritychange_event/index.md (hunk 6 above)
Comment (teoli2003) [reply files/en-us/web/api/prioritychange_event/index.md:2]: Note that as there are spaces in the title, you need the ticks.

File: files/en-us/web/api/prioritychange_event/index.md (hunk 7 above)
Comment (hamishwillee) [reply files/en-us/web/api/prioritychange_event/index.md:3]: Thanks @teoli2003 - really appreciate your help on this.

File: files/en-us/web/api/taskprioritychangeevent/taskprioritychangeevent/index.md (hunk 1 above)
Comment (hamishwillee) [reply files/en-us/web/api/taskprioritychangeevent/taskprioritychangeevent/index.md:32]: Thanks very much. Will fix.

File: files/en-us/web/api/scheduler/posttask/index.md [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+{{APIRef("Prioritized Task Scheduling API")}} {{SeeCompatTable}}
+
+The **`postTask()`** method of the {{domxref("Scheduler")}} interface is used for adding prioritized tasks to be [scheduled](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API).
+
+The method allows you to specify a minimum delay before the task will run, a priority for the task, and a signal that can be used to abort the task.
+
+## Syntax
+
+```js
+postTask(callback, options)
```
Comment (teoli2003) [inline files/en-us/web/api/scheduler/posttask/index.md:21]: Drive-by comment.

If the `options` parameter is optional, as indicated below – I didn't check if this was indeed the case –, we should have two lines here.
```suggestion
postTask(callback)
postTask(callback, options)
```

Comment (hamishwillee) [reply files/en-us/web/api/scheduler/posttask/index.md:21]: Thanks very much. I haven't done this doc yet. I'll ping you when this is is all ready for more review. I want to write some code first and finish what I can get out of the specs and explainer.

File: files/en-us/web/api/prioritized_task_scheduling_api/index.md [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+  - : Interface for a signal object that allows you to abort a task and change its priority, if required, using a {{domxref("TaskController")}} object.
+- {{domxref("TaskPriorityChangeEvent")}}
+  - : The interface for the {{domxref("TaskSignal/prioritychange_event","prioritychange")}} event, sent when the priority for a task is changed.
+
+> **Note:** If the [task priority](#task_priorities) never needs to be changed, you can use an {{domxref("AbortController")}} and its associated {{domxref("AbortSignal")}} instead of {{domxref("TaskController")}} and {{domxref("TaskSignal")}}.
+
+
+### Extensions to other interfaces
+
+The Priorizied Task Scheduling API extends the following APIs, adding the listed features:
```
Comment (hamishwillee) [inline files/en-us/web/api/prioritized_task_scheduling_api/index.md:134]: @teoli2003 I'm not sure this is what is meant by "extension to other interfaces". What I have done is added the `window.scheduler` - because this was added to window so you can get a handle to the API. Is this what is meant?

2. Below you see `Worker.scheduler`

The spec actually adds this to [WindowOrWorderGlobalScope](https://wicg.github.io/scheduling-apis/#sec-patches-html-windoworworkerglobalscope) mixin which we don't document. Given that, where do I add the worker entries. I also asked this question in BCD: https://github.com/mdn/browser-compat-data/pull/16409

This is the only missing doc

**Comment by hamishwillee** [comment]
@teoli2003 This is getting close but isn't quite ready for _full review_. However I would appreciate a partial review of the bits that have been done if you are able? 

In particular
- the top level API doc,  
- `TaskController` (top level, not methods) 
-  `TaskPriorityChangeEvent` (and the `prioritychange_event`),
- `Windows.scheduler`

Still need to look at examples for the rest. Mostly I'm after a sanity check that it is at the right level and I'm not using patterns we hate.

File: files/en-us/web/api/window/scheduler/index.md [hunk 12]
```diff
@@ -0,0 +1,66 @@
+---
```
Comment (hamishwillee) [inline files/en-us/web/api/window/scheduler/index.md:1]: FYI , this is the global object. To prevent file clash with the `Scheduler` interface this file lives under Window. However it is described as a global object where it is referenced.

**Comment by hamishwillee** [comment]
This is now ready for review. It is waiting on BCD for the global scheduler property, and I will also need to add that to the sidebar, provided this gets merged.

File: files/en-us/mozilla/firefox/experimental_features/index.md [hunk 13]
```diff
@@ -1528,6 +1528,15 @@ Note that since locking the screen orientation isn't typically supported on desk
   </tbody>
 </table>
 
+
+### Prioritized Task Scheduling API
+
+The [Prioritized Task Scheduling API](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API) provides a standardized way to prioritize all tasks belonging to an application, whether they defined in a website developer's code, or in third party libraries and frameworks.
+
+This is enabled on Firefox Nightly (only) from Firefox 101.
+No preference is provided to allow it to be enabled in other releases.
```
Comment (hamishwillee) [inline files/en-us/mozilla/firefox/experimental_features/index.md:1537]: As no preference provided you can't turn is on for other releases. As such, no point in adding the normal table showing what's enabled or not in each release.

File: files/en-us/web/api/scheduler/index.md [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+## Properties
+
+None.
+
+## Methods
+
+- {{domxref('Scheduler.postTask()')}}
+  - : Adds a task to the scheduler as a callback, optionally specifying a priority, delay, and/or a signal for aborting the task.
+    
+
```
Comment (teoli2003) [inline files/en-us/web/api/scheduler/index.md:27]: NIT, drive-by comment: an extra line (Onkar's daily script would catch this anyway)
```suggestion
```

File: files/en-us/web/api/scheduler/posttask/index.md [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+      - : The minimum amount of time after which the task will be added to the scheduler queue, in whole milliseconds.
+        The actual delay may be higher than specified, but will not be less.
+        The default delay is 0.
+
+### Return Value
+
+Returns a {{jsxref("Promise")}} that is resolved with the return value of the `callback` function, or which may be rejected with the `signal`'s abort reason ({{domxref("AbortSignal.reason")}}).
+The promise may also be rejected with an error thrown by the callback during execution.
+
+
```
Comment (teoli2003) [inline files/en-us/web/api/scheduler/posttask/index.md:70]: Extra line
```suggestion
```

File: files/en-us/web/api/scheduler/posttask/index.md [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+// Check that feature is supported
+if ('scheduler' in this) {
+  console.log('Feature: Supported');
+}
+else {
+  console.log('Feature: NOT Supported');
+}
+```
+
+
```
Comment (teoli2003) [inline files/en-us/web/api/scheduler/posttask/index.md:91]: Extra line
```suggestion
```

File: files/en-us/web/api/taskcontroller/index.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+```
+
+The console output of this example would be:
+```
+The priority changed from user-blocking to background.
+Task aborted
+Catch error: AbortError
+```
+
+
```
Comment (teoli2003) [inline files/en-us/web/api/taskcontroller/index.md:124]: Extra line
```suggestion
```

File: files/en-us/web/api/taskcontroller/setpriority/index.md [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+None ({{jsxref("undefined")}}).
+
+### Exceptions
+
+- `NotAllowedError` {{domxref("DOMException")}}
+  - : A priority change has been started when one is already running.
+
+## Examples
+
+
```
Comment (teoli2003) [inline files/en-us/web/api/taskcontroller/setpriority/index.md:47]:
```suggestion
```

File: files/en-us/web/api/taskcontroller/setpriority/index.md [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+The controller can then be used to change the priority
+
+```js
+// Change the priority to 'background' using the controller
+controller.setPriority('background');
+```
+
+Additional examples, including showing how to handle the event that results from changing the priority, can be found in: [Prioritized Task Scheduling API > Examples](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#examples).
+
+
```
Comment (teoli2003) [inline files/en-us/web/api/taskcontroller/setpriority/index.md:75]:
```suggestion
```

File: files/en-us/web/api/taskcontroller/taskcontroller/index.md [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+### Parameters
+
+- `init` {{optional_inline}}
+  - : An object with the following properties:
+    
+    - `priority` {{optional_inline}}
+      - : The [priority](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#task_priorities) of the signal associated with this `TaskController`.
+        One of: `"user-blocking"`, `"user-visible"` (default), `"background"`.
+
+
```
Comment (teoli2003) [inline files/en-us/web/api/taskcontroller/taskcontroller/index.md:35]:
```suggestion
```

File: files/en-us/web/api/tasksignal/prioritychange_event/index.md [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+{{InheritanceDiagram("TaskPriorityChangeEvent")}}
+
+
+## Event properties
+
+- {{domxref('TaskPriorityChangeEvent.previousPriority')}}
+  - : Indicates the previous priority of the task (before it was changed).
+    The new/updated priority is read from `event.target.priority` ([`TaskSignal.priority`](/en-US/docs/Web/API/TaskSignal/priority)).
+
+
```
Comment (teoli2003) [inline files/en-us/web/api/tasksignal/prioritychange_event/index.md:40]:
```suggestion
```

File: files/en-us/web/api/tasksignal/prioritychange_event/index.md [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+> **Note:** The code above uses a custom logging function `mylog()` to log to the text area below.
+> This is hidden as it isn't relevant to the example.
+
+The output below demonstrates shows that the [task's priority](/en-US/docs/Web/API/Prioritized_Task_Scheduling_API#task_priorities) changed from `user-blocking` to `background`.
+This happens before the task is executed, but could also happen when the task is running.
+
+{{EmbedLiveSample("Examples",'400px','130px')}}
+
+
```
Comment (teoli2003) [inline files/en-us/web/api/tasksignal/prioritychange_event/index.md:88]:
```suggestion
```

File: files/en-us/web/api/workerglobalscope/index.md [hunk 23]
```diff
@@ -30,6 +30,7 @@ _This interface inherits properties from the {{domxref("EventTarget")}} interfac
 - {{domxref("WorkerGlobalScope.location")}} {{readOnlyinline}}
   - : Returns the {{domxref("WorkerLocation")}} associated with the worker. It is a specific location object, mostly a subset of the {{domxref("Location")}} for browsing scopes, but adapted to workers.
 
+
```
Comment (teoli2003) [inline files/en-us/web/api/workerglobalscope/index.md:33]:
```suggestion
```

**Review by teoli2003 — APPROVED** [review]
👍

**Comment by teoli2003** [comment]
@hamishwillee Not sure if you want more people to review, so I'm not merging. But if not, you can just merge it.

**Comment by hamishwillee** [comment]
Thanks very much @teoli2003 . I wasn't sure if you'd done a drive-by review or a full review - so I'd also asked Will for a review. Since you're happy and he's on holiday, let's merge. We can always iterated (it is unlikely to be "wrong", just not necessarily best examples etc.)
